{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading in data available through The Arabidopsis Information Resource (TAIR) database<br>\n",
    "The purpose of this notebook is to read in and do a preliminary survey of the data related to text and ontology annotations of text that is available through TAIR. The datasets need to be organized and also restructured into a standard format that will allow it be combined with datasets from other resources. This notebook takes the following input files that were downloaded from the TAIR and produces a set of files that have standard columns, which are listed and described below.<br>\n",
    "<br>\n",
    "### Files read<br>\n",
    "```<br>\n",
    "plant-data/databases/tair/Locus_Germplasm_Phenotype_20190930.txt<br>\n",
    "plant-data/databases/tair/<br>\n",
    "plant-data/databases/tair/po_anatomy_gene_arabidopsis_tair.assoc<br>\n",
    "plant-data/databases/tair/po_temporal_gene_arabidopsis_tair.assoc<br>\n",
    "plant-data/databases/tair/ATH_GO_GOSLIM.txt<br>\n",
    "```<br>\n",
    "<br>\n",
    "### Files created<br>\n",
    "```<br>\n",
    "plant-data/reshaped_data/tair_phenotype_descriptions.csv<br>\n",
    "plant-data/reshaped_data/tair_general_descriptions.csv<br>\n",
    "plant-data/reshaped_data/tair_all_go_annotations.csv<br>\n",
    "plant-data/reshaped_data/tair_curated_go_annotations.csv<br>\n",
    "plant-data/reshaped_data/tair_curated_po_annotations.csv<br>\n",
    "```<br>\n",
    "<br>\n",
    "### Columns in the created files<br>\n",
    "* **species**: A string indicating what species the gene is in, currently uses the 3-letter codes from the KEGG database.<br>\n",
    "* **unique_gene_identifiers**: Pipe delimited list of gene identifiers, names, models, etc which must uniquely refer to this gene.<br>\n",
    "* **other_gene_identifiers**: Pipe delimited list of other identifiers, names, aliases, synonyms for the gene, which may but do not have to uniquely refer to it.<br>\n",
    "* **gene_models**: Pipe delimited list of gene model names that map to this gene.<br>\n",
    "* **descriptions**: A free text field for any descriptions of phenotypes associated with this gene.<br>\n",
    "* **annotations**: Pipe delimited list of gene ontology term identifiers.<br>\n",
    "* **sources**: Pipe delimited list of strings that indicate where this data comes from such as database names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.tokenize import sent_tokenize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append(\"../utils\")\n",
    "from constants import NCBI_TAG, EVIDENCE_CODES, ABBREVIATIONS_MAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append(\"../../oats\")\n",
    "from oats.nlp.preprocess import concatenate_with_delim, subtract_string_lists, replace_delimiter, concatenate_texts\n",
    "from oats.nlp.small import remove_punctuation, remove_enclosing_brackets, add_prefix_safely"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUTPUT_DIR = \"../reshaped_data\"\n",
    "mpl.rcParams[\"figure.dpi\"] = 200\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Columns that should be in the final reshaped files.\n",
    "reshaped_columns = [\"species\", \n",
    " \"unique_gene_identifiers\", \n",
    " \"other_gene_identifiers\", \n",
    " \"gene_models\", \n",
    " \"descriptions\", \n",
    " \"annotations\", \n",
    " \"sources\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating and testing a lambda for finding gene model strings.\n",
    "gene_model_pattern = re.compile(\"at[0-9]{1}g[0-9]+\")\n",
    "is_gene_model = lambda s: bool(gene_model_pattern.match(s.lower()))\n",
    "assert is_gene_model(\"AT1G34534\") == True\n",
    "assert is_gene_model(\"at2g23452\") == True\n",
    "assert is_gene_model(\"ACAB1\") == False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with genes and phenotype descriptions (Locus_Germplasm_Phenotype_20180702.txt)<br>\n",
    "Reading in the dataset of phenotypic descriptions. There is only one value specified as the gene name (locus name) in the original dataset so this column does not need to be parsed further. The descriptions commonly use semi-colons to separate phrases. The next cell gets the distribution of the number of phrases in each description field for the dataset of text descriptions, as determined by a sentence parser. The majority of the descriptions are a single sentence or phrase, but some contain more."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9225</th>\n",
       "      <td>AT4G12810</td>\n",
       "      <td>Resemble BR deficient mutants. Short hypocotyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9340</th>\n",
       "      <td>AT4G14960</td>\n",
       "      <td>Right-handed helical growth in the root and ot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13742</th>\n",
       "      <td>AT5G64740</td>\n",
       "      <td>short, thick root and dark-grown hypocotyl; no...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8313</th>\n",
       "      <td>AT3G54810</td>\n",
       "      <td>Reduced germination rate. Increased dormancy.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12203</th>\n",
       "      <td>AT5G24270</td>\n",
       "      <td>Decrease of macronutrients concentration in th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13158</th>\n",
       "      <td>AT5G54260</td>\n",
       "      <td>Hyperaccumulation of  AtSPO11-1 in meiocytes.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15379</th>\n",
       "      <td>EMB274</td>\n",
       "      <td>developmental arrest of mutant embryos occurs ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2677</th>\n",
       "      <td>AT1G64060</td>\n",
       "      <td>Four-weeks old plants have necrotic lesions an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11416</th>\n",
       "      <td>AT5G11150</td>\n",
       "      <td>The mutant plants are indistinguishable from t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11331</th>\n",
       "      <td>AT5G10140</td>\n",
       "      <td>not late flowering alone; when combined with F...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5910</th>\n",
       "      <td>AT2G44990</td>\n",
       "      <td>bushy appearance due to increased axillary bra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12405</th>\n",
       "      <td>AT5G35620</td>\n",
       "      <td>Immunity to potyviruses</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14908</th>\n",
       "      <td>EMB2062</td>\n",
       "      <td>embryo defective; terminal embryo phenotype: p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3240</th>\n",
       "      <td>AT1G72560</td>\n",
       "      <td>Leaf initiation is delayed slightly after germ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5871</th>\n",
       "      <td>AT2G44830</td>\n",
       "      <td>Shorter root length, reduced primary root grow...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7263</th>\n",
       "      <td>AT3G23150</td>\n",
       "      <td>Ethylene insensitivity reflected in etiolated ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1775</th>\n",
       "      <td>AT1G30520</td>\n",
       "      <td>seedling lethal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2782</th>\n",
       "      <td>AT1G65620</td>\n",
       "      <td>The rosette leaf petioles of the triple mutant...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14680</th>\n",
       "      <td>EMB1733</td>\n",
       "      <td>embryo defective; terminal embryo phenotype: p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8609</th>\n",
       "      <td>AT3G61510</td>\n",
       "      <td>no dramatic change in phenotype</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      unique_gene_identifiers                                       descriptions\n",
       "9225                AT4G12810  Resemble BR deficient mutants. Short hypocotyl...\n",
       "9340                AT4G14960  Right-handed helical growth in the root and ot...\n",
       "13742               AT5G64740  short, thick root and dark-grown hypocotyl; no...\n",
       "8313                AT3G54810      Reduced germination rate. Increased dormancy.\n",
       "12203               AT5G24270  Decrease of macronutrients concentration in th...\n",
       "13158               AT5G54260      Hyperaccumulation of  AtSPO11-1 in meiocytes.\n",
       "15379                  EMB274  developmental arrest of mutant embryos occurs ...\n",
       "2677                AT1G64060  Four-weeks old plants have necrotic lesions an...\n",
       "11416               AT5G11150  The mutant plants are indistinguishable from t...\n",
       "11331               AT5G10140  not late flowering alone; when combined with F...\n",
       "5910                AT2G44990  bushy appearance due to increased axillary bra...\n",
       "12405               AT5G35620                            Immunity to potyviruses\n",
       "14908                 EMB2062  embryo defective; terminal embryo phenotype: p...\n",
       "3240                AT1G72560  Leaf initiation is delayed slightly after germ...\n",
       "5871                AT2G44830  Shorter root length, reduced primary root grow...\n",
       "7263                AT3G23150  Ethylene insensitivity reflected in etiolated ...\n",
       "1775                AT1G30520                                    seedling lethal\n",
       "2782                AT1G65620  The rosette leaf petioles of the triple mutant...\n",
       "14680                 EMB1733  embryo defective; terminal embryo phenotype: p...\n",
       "8609                AT3G61510                    no dramatic change in phenotype"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"../databases/tair/Locus_Germplasm_Phenotype_20190930.txt\"\n",
    "usecols = [\"LOCUS_NAME\", \"PHENOTYPE\"]\n",
    "usenames = [\"unique_gene_identifiers\", \"descriptions\"]\n",
    "renamed = {k:v for k,v in zip(usecols,usenames)}\n",
    "df = pd.read_table(filename, usecols=usecols)\n",
    "df.rename(columns=renamed, inplace=True)\n",
    "df.dropna(axis=\"rows\",inplace=True)\n",
    "df.sample(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15442, 2)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plotting distributions of number of phrases in each description."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plotting distributions of number of phrases in each description.\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "ax1.set_title(\"Phenotype Descriptions\")\n",
    "ax2.set_title(\"Phenotype Descriptions\")\n",
    "ax1.set_xlabel(\"Number of phrases\")\n",
    "ax2.set_xlabel(\"Number of words\")\n",
    "x1 = [len(sent_tokenize(x)) for x in df[\"descriptions\"].values]\n",
    "x2 = [len(word_tokenize(x)) for x in df[\"descriptions\"].values]\n",
    "ax1.hist(x1, bins=15, range=(0,15), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "ax2.hist(x2, bins=30, range=(0,200), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "fig.set_size_inches(15,4)\n",
    "fig.tight_layout()\n",
    "fig.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Restructuring the dataset to include all expected column names.\n",
    "df[\"species\"] = \"ath\"\n",
    "df[\"other_gene_identifiers\"] = \"\"\n",
    "df[\"gene_models\"] = df[\"unique_gene_identifiers\"].map(lambda x: \"|\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df[\"annotations\"] = \"\"\n",
    "df[\"sources\"] = \"TAIR\"\n",
    "df = df[reshaped_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>other_gene_identifiers</th>\n",
       "      <th>gene_models</th>\n",
       "      <th>descriptions</th>\n",
       "      <th>annotations</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>ACL2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>flower stems are much reduced in length, semi-...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>ADE1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sustained and enhanced levels of ABA-regulated...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALB2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>white embryo and seedling (albino), lethal.</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALR-104</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>incomplete penetrance; increased aluminum resi...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALR-128</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>incomplete penetrance; increased aluminum resi...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALS4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>increased aluminum sensitivity; poor root grow...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALS5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>incomplete penetrance; increased aluminum sens...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC10</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no visible whole plant phenotypic effect; mean...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mesophyll cells contain approximately half the...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>fewer and larger chloroplasts than wild type i...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC7</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>pale first leaves in early seedling growth; me...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC8</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no visible whole plant phenotypic effect;  mea...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no visible whole plant phenotypic effect; mean...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARS1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Semi-dominant arsenic resistance phenotype. In...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; globular</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; globular</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species unique_gene_identifiers other_gene_identifiers gene_models                                       descriptions annotations sources\n",
       "0      ath                    ACL2                                     flower stems are much reduced in length, semi-...                TAIR\n",
       "1      ath                    ADE1                                     sustained and enhanced levels of ABA-regulated...                TAIR\n",
       "2      ath                    ALB2                                           white embryo and seedling (albino), lethal.                TAIR\n",
       "3      ath                 ALR-104                                     incomplete penetrance; increased aluminum resi...                TAIR\n",
       "4      ath                 ALR-128                                     incomplete penetrance; increased aluminum resi...                TAIR\n",
       "5      ath                    ALS4                                     increased aluminum sensitivity; poor root grow...                TAIR\n",
       "6      ath                    ALS5                                     incomplete penetrance; increased aluminum sens...                TAIR\n",
       "7      ath                   ARC10                                     no visible whole plant phenotypic effect; mean...                TAIR\n",
       "8      ath                    ARC2                                     mesophyll cells contain approximately half the...                TAIR\n",
       "9      ath                    ARC4                                     fewer and larger chloroplasts than wild type i...                TAIR\n",
       "10     ath                    ARC7                                     pale first leaves in early seedling growth; me...                TAIR\n",
       "11     ath                    ARC8                                     no visible whole plant phenotypic effect;  mea...                TAIR\n",
       "12     ath                    ARC9                                     no visible whole plant phenotypic effect; mean...                TAIR\n",
       "13     ath                    ARS1                                     Semi-dominant arsenic resistance phenotype. In...                TAIR\n",
       "14     ath               AT1G01040                          AT1G01040  embryo defective; developmental arrest of muta...                TAIR\n",
       "15     ath               AT1G01040                          AT1G01040  embryo defective; developmental arrest of muta...                TAIR\n",
       "16     ath               AT1G01040                          AT1G01040  embryo defective; developmental arrest of muta...                TAIR\n",
       "17     ath               AT1G01040                          AT1G01040  embryo defective; developmental arrest of muta...                TAIR\n",
       "18     ath               AT1G01040                          AT1G01040                         embryo defective; globular                TAIR\n",
       "19     ath               AT1G01040                          AT1G01040                         embryo defective; globular                TAIR"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Outputting the dataset of phenotype descriptions to csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_phenotype_descriptions.csv\")\n",
    "df.to_csv(path, index=False)\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with curator summaries (Araport11_functional_descriptions_20190930.txt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6444</th>\n",
       "      <td>AT2G38470.1</td>\n",
       "      <td>Member of the plant WRKY transcription factor ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24886</th>\n",
       "      <td>AT4G21940.1</td>\n",
       "      <td>member of Calcium Dependent Protein Kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3472</th>\n",
       "      <td>AT1G31820.1</td>\n",
       "      <td>Encodes POLYAMINE UPTAKE TRANSPORTER 1, an ami...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32730</th>\n",
       "      <td>AT4G19230.2</td>\n",
       "      <td>Encodes a protein with ABA 8'-hydroxylase acti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24143</th>\n",
       "      <td>AT2G21040.1</td>\n",
       "      <td>C2 domain-containing protein. Possible pseudog...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12288</th>\n",
       "      <td>AT3G57530.1</td>\n",
       "      <td>Calcium-dependent Protein Kinase. ABA signalin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14089</th>\n",
       "      <td>AT4G13930.1</td>\n",
       "      <td>Encodes a serine hydroxymethyltransferase maxi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>RPP2</td>\n",
       "      <td>Confers resistance to Peronospora parasitica. ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56851</th>\n",
       "      <td>AT4G37900.2</td>\n",
       "      <td>Protein of unknown function that contains DUF1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49105</th>\n",
       "      <td>AT5G02320.3</td>\n",
       "      <td>Encodes a putative c-MYB-like transcription fa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61652</th>\n",
       "      <td>AT5G14400.2</td>\n",
       "      <td>Encodes a brassinosteroid C-22 hydroxylase.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>EMB224</td>\n",
       "      <td>a genetic locus involved in embryogenesis. Mut...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36913</th>\n",
       "      <td>AT3G26816.1</td>\n",
       "      <td>Encodes a microRNA that targets several HAP2 f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26324</th>\n",
       "      <td>AT2G31170.1</td>\n",
       "      <td>Encodes the cysteinyl t-RNA synthetase SYCO AR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55472</th>\n",
       "      <td>AT5G66030.4</td>\n",
       "      <td>Involved in golgi protein trafficking. AtARL1 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5421</th>\n",
       "      <td>AT2G21410.1</td>\n",
       "      <td>Vacuolar proton ATPase subunit VHA-a isoform 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18193</th>\n",
       "      <td>AT5G39670.1</td>\n",
       "      <td>Calmodulin like protein involved in negative r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47715</th>\n",
       "      <td>AT1G26700.2</td>\n",
       "      <td>A member of a large family of seven-transmembr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46266</th>\n",
       "      <td>AT4G01540.4</td>\n",
       "      <td>Encodes a membrane-bound NAC (for NAM, ATAF1/2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35027</th>\n",
       "      <td>AT4G31600.2</td>\n",
       "      <td>Encodes a Golgi-localized UDP–glucose/UDP–gala...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      unique_gene_identifiers                                       descriptions\n",
       "6444              AT2G38470.1  Member of the plant WRKY transcription factor ...\n",
       "24886             AT4G21940.1         member of Calcium Dependent Protein Kinase\n",
       "3472              AT1G31820.1  Encodes POLYAMINE UPTAKE TRANSPORTER 1, an ami...\n",
       "32730             AT4G19230.2  Encodes a protein with ABA 8'-hydroxylase acti...\n",
       "24143             AT2G21040.1  C2 domain-containing protein. Possible pseudog...\n",
       "12288             AT3G57530.1  Calcium-dependent Protein Kinase. ABA signalin...\n",
       "14089             AT4G13930.1  Encodes a serine hydroxymethyltransferase maxi...\n",
       "228                      RPP2  Confers resistance to Peronospora parasitica. ...\n",
       "56851             AT4G37900.2  Protein of unknown function that contains DUF1...\n",
       "49105             AT5G02320.3  Encodes a putative c-MYB-like transcription fa...\n",
       "61652             AT5G14400.2        Encodes a brassinosteroid C-22 hydroxylase.\n",
       "103                    EMB224  a genetic locus involved in embryogenesis. Mut...\n",
       "36913             AT3G26816.1  Encodes a microRNA that targets several HAP2 f...\n",
       "26324             AT2G31170.1  Encodes the cysteinyl t-RNA synthetase SYCO AR...\n",
       "55472             AT5G66030.4  Involved in golgi protein trafficking. AtARL1 ...\n",
       "5421              AT2G21410.1  Vacuolar proton ATPase subunit VHA-a isoform 2...\n",
       "18193             AT5G39670.1  Calmodulin like protein involved in negative r...\n",
       "47715             AT1G26700.2  A member of a large family of seven-transmembr...\n",
       "46266             AT4G01540.4  Encodes a membrane-bound NAC (for NAM, ATAF1/2...\n",
       "35027             AT4G31600.2  Encodes a Golgi-localized UDP–glucose/UDP–gala..."
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"../databases/tair/Araport11_functional_descriptions_20190930.txt\"\n",
    "usecols = [\"name\", \"Curator_summary\"]\n",
    "usenames = [\"unique_gene_identifiers\", \"descriptions\"]\n",
    "renamed = {k:v for k,v in zip(usecols,usenames)}\n",
    "df = pd.read_table(filename, usecols=usecols)\n",
    "df.rename(columns=renamed, inplace=True)\n",
    "df.dropna(axis=\"rows\",inplace=True)\n",
    "df.sample(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20630, 2)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Restructuring the dataset to include all expected column names.\n",
    "df[\"species\"] = \"ath\"\n",
    "df[\"other_gene_identifiers\"] = \"\"\n",
    "df[\"gene_models\"] = df[\"unique_gene_identifiers\"].map(lambda x: \"|\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df[\"annotations\"] = \"\"\n",
    "df[\"sources\"] = \"TAIR\"\n",
    "df = df[reshaped_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>other_gene_identifiers</th>\n",
       "      <th>gene_models</th>\n",
       "      <th>descriptions</th>\n",
       "      <th>annotations</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60016</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT3G20530.2</td>\n",
       "      <td></td>\n",
       "      <td>AT3G20530.2</td>\n",
       "      <td>Protein kinase superfamily protein, expressed ...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38592</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT3G22170.2</td>\n",
       "      <td></td>\n",
       "      <td>AT3G22170.2</td>\n",
       "      <td>A component of the PHYA signaling network, med...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58701</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G53120.7</td>\n",
       "      <td></td>\n",
       "      <td>AT5G53120.7</td>\n",
       "      <td>encodes a novel spermine synthase and is a par...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45084</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G25370.3</td>\n",
       "      <td></td>\n",
       "      <td>AT5G25370.3</td>\n",
       "      <td>member of C2-PLD subfamily. Analyses on the ge...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7445</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G38700.1</td>\n",
       "      <td></td>\n",
       "      <td>AT2G38700.1</td>\n",
       "      <td>Encodes mevalonate diphosphate decarboxylase, ...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35444</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G19390.4</td>\n",
       "      <td></td>\n",
       "      <td>AT5G19390.4</td>\n",
       "      <td>Encodes a protein with similarity to REN1, a R...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31581</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALI</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>belongs to the DIS(distorted) gene family. Aff...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37123</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT4G08390.3</td>\n",
       "      <td></td>\n",
       "      <td>AT4G08390.3</td>\n",
       "      <td>Encodes a chloroplastic stromal ascorbate pero...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45893</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G50375.3</td>\n",
       "      <td></td>\n",
       "      <td>AT5G50375.3</td>\n",
       "      <td>Converts pentacyclic cyclopropyl sterols to co...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17570</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G38530.1</td>\n",
       "      <td></td>\n",
       "      <td>AT5G38530.1</td>\n",
       "      <td>TSBtype2 encodes a type 2 tryptophan synthase ...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20504</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G14740.1</td>\n",
       "      <td></td>\n",
       "      <td>AT5G14740.1</td>\n",
       "      <td>Encodes a beta carbonic anhydrase likely to be...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2207</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G07540.1</td>\n",
       "      <td></td>\n",
       "      <td>AT1G07540.1</td>\n",
       "      <td>Arabidopsis thaliana telomere-binding protein,...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19623</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G56290.1</td>\n",
       "      <td></td>\n",
       "      <td>AT5G56290.1</td>\n",
       "      <td>Encodes the peroxisomal targeting signal type ...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34291</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G64195.1</td>\n",
       "      <td></td>\n",
       "      <td>AT1G64195.1</td>\n",
       "      <td>Encodes a defensin-like (DEFL) family protein.</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37991</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G48360.2</td>\n",
       "      <td></td>\n",
       "      <td>AT1G48360.2</td>\n",
       "      <td>Encodes a FAN1 homolog that is involved in int...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34284</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G35537.1</td>\n",
       "      <td></td>\n",
       "      <td>AT1G35537.1</td>\n",
       "      <td>Encodes a defensin-like peptide with antifunga...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13055</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT4G21990.1</td>\n",
       "      <td></td>\n",
       "      <td>AT4G21990.1</td>\n",
       "      <td>Encodes a protein disulfide isomerase-like (PD...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49973</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT4G23180.2</td>\n",
       "      <td></td>\n",
       "      <td>AT4G23180.2</td>\n",
       "      <td>Encodes a receptor-like protein kinase. Naming...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24383</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G40330.1</td>\n",
       "      <td></td>\n",
       "      <td>AT2G40330.1</td>\n",
       "      <td>Encodes a member of the PYR (pyrabactin resist...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26663</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT3G09940.1</td>\n",
       "      <td></td>\n",
       "      <td>AT3G09940.1</td>\n",
       "      <td>Encodes a member of the monodehydroascorbate r...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      species unique_gene_identifiers other_gene_identifiers  gene_models                                       descriptions annotations sources\n",
       "60016     ath             AT3G20530.2                         AT3G20530.2  Protein kinase superfamily protein, expressed ...                TAIR\n",
       "38592     ath             AT3G22170.2                         AT3G22170.2  A component of the PHYA signaling network, med...                TAIR\n",
       "58701     ath             AT5G53120.7                         AT5G53120.7  encodes a novel spermine synthase and is a par...                TAIR\n",
       "45084     ath             AT5G25370.3                         AT5G25370.3  member of C2-PLD subfamily. Analyses on the ge...                TAIR\n",
       "7445      ath             AT2G38700.1                         AT2G38700.1  Encodes mevalonate diphosphate decarboxylase, ...                TAIR\n",
       "35444     ath             AT5G19390.4                         AT5G19390.4  Encodes a protein with similarity to REN1, a R...                TAIR\n",
       "31581     ath                     ALI                                      belongs to the DIS(distorted) gene family. Aff...                TAIR\n",
       "37123     ath             AT4G08390.3                         AT4G08390.3  Encodes a chloroplastic stromal ascorbate pero...                TAIR\n",
       "45893     ath             AT5G50375.3                         AT5G50375.3  Converts pentacyclic cyclopropyl sterols to co...                TAIR\n",
       "17570     ath             AT5G38530.1                         AT5G38530.1  TSBtype2 encodes a type 2 tryptophan synthase ...                TAIR\n",
       "20504     ath             AT5G14740.1                         AT5G14740.1  Encodes a beta carbonic anhydrase likely to be...                TAIR\n",
       "2207      ath             AT1G07540.1                         AT1G07540.1  Arabidopsis thaliana telomere-binding protein,...                TAIR\n",
       "19623     ath             AT5G56290.1                         AT5G56290.1  Encodes the peroxisomal targeting signal type ...                TAIR\n",
       "34291     ath             AT1G64195.1                         AT1G64195.1     Encodes a defensin-like (DEFL) family protein.                TAIR\n",
       "37991     ath             AT1G48360.2                         AT1G48360.2  Encodes a FAN1 homolog that is involved in int...                TAIR\n",
       "34284     ath             AT1G35537.1                         AT1G35537.1  Encodes a defensin-like peptide with antifunga...                TAIR\n",
       "13055     ath             AT4G21990.1                         AT4G21990.1  Encodes a protein disulfide isomerase-like (PD...                TAIR\n",
       "49973     ath             AT4G23180.2                         AT4G23180.2  Encodes a receptor-like protein kinase. Naming...                TAIR\n",
       "24383     ath             AT2G40330.1                         AT2G40330.1  Encodes a member of the PYR (pyrabactin resist...                TAIR\n",
       "26663     ath             AT3G09940.1                         AT3G09940.1  Encodes a member of the monodehydroascorbate r...                TAIR"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Outputting the dataset of phenotype descriptions to csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_general_descriptions.csv\")\n",
    "df.to_csv(path, index=False)\n",
    "df.sample(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with gene ontology annotations (ATH_GO_GOSLIM.txt)<br>\n",
    "Read in the file containing names of loci and corresponding information relating to gene ontology term annotation. Not all of the columns are used here, only a subset of them are read in. The relationship column refers to the relationships between the gene for that loci and the term mentioned on that given line. Evidence refer to the method of acquiring and the confidence in the annotation itself. This is retained so that we can subset that dataset based on whether the annotations are experimentally confirmed or simply predicted annotations. This section also looks at how many unique values are present for each field.<br>\n",
    "<br>\n",
    "Each term annotation in this dataset is also associated with an evidence code specifying the method by which this annotation was made, which is related to the confidence that we can have in this annotation, and the tasks that the annotation should be used for. About half of the term annotations were made computationally, but there are also a high number of annotations available from high confidence annotations such as experimentally validated, curator statements, and author statements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       locus       object     term_id evidence_code                              reference\n",
      "0  AT1G01010    AT1G01010  GO:0006355           IEA            AnalysisReference:501756966\n",
      "1  AT1G01010    AT1G01010  GO:0006355           ISS    Publication:1345963|PMID:11118137  \n",
      "2  AT1G01010    AT1G01010  GO:0044212           IPI  Publication:501786139|PMID:30356219  \n",
      "3  AT1G01010    AT1G01010  GO:0005634           IEA            AnalysisReference:501756971\n",
      "4  AT1G01010    AT1G01010  GO:0006355           IEA            AnalysisReference:501756966\n",
      "5  AT1G01010    AT1G01010  GO:0006355           IEA            AnalysisReference:501756966\n",
      "6  AT1G01010    AT1G01010  GO:0006355           IEA            AnalysisReference:501756966\n",
      "7  AT1G01010  AT1G01010.1  GO:0005634           ISM            AnalysisReference:501780126\n",
      "8  AT1G01010    AT1G01010  GO:0006355           ISS    Publication:1345963|PMID:11118137  \n",
      "9  AT1G01010    AT1G01010  GO:0003700           ISS    Publication:1345963|PMID:11118137  \n",
      "(343166, 7)\n",
      "locus                31302\n",
      "object               68581\n",
      "relationship            19\n",
      "term_label            7317\n",
      "term_id               7317\n",
      "evidence_code           16\n",
      "reference            12158\n"
     ]
    }
   ],
   "source": [
    "filename = \"../databases/tair/ATH_GO_GOSLIM.txt\"\n",
    "df_go = pd.read_table(filename, header=None, usecols=[0,2,3,4,5,9,12])\n",
    "df_go.columns = [\"locus\",\"object\",\"relationship\",\"term_label\",\"term_id\",\"evidence_code\",\"reference\"]\n",
    "unique_values = {col:len(pd.unique(df_go[col].values)) for col in df_go.columns}\n",
    "print(df_go[[\"locus\",\"object\",\"term_id\",\"evidence_code\",\"reference\"]].head(10))\n",
    "print(df_go.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "experimental               131557\n",
      "curator_statement           22809\n",
      "author_statement            17092\n",
      "phylogenetics               60627\n",
      "high_throughput              5914\n",
      "computational               47057\n",
      "electronic_annotation       58110\n"
     ]
    }
   ],
   "source": [
    "code_quantities = {c:len([x for x in df_go[\"evidence_code\"] if EVIDENCE_CODES[x] in c]) \n",
    "             for c in list(set(EVIDENCE_CODES.values()))}\n",
    "for k,v in code_quantities.items():\n",
    "    print(\"{:25}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Restructuring the dataset to include all the expected column names.\n",
    "df_go[\"species\"] = \"ath\"\n",
    "df_go[\"unique_gene_identifiers\"] = df_go[\"locus\"]\n",
    "df_go[\"other_gene_identifiers\"] = \"\"\n",
    "df_go[\"gene_models\"] = df_go[\"unique_gene_identifiers\"].map(lambda x: \"\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df_go[\"descriptions\"] = \"\"\n",
    "df_go[\"annotations\"] = df_go[\"term_id\"]\n",
    "df_go[\"sources\"] = \"TAIR\"\n",
    "high_confidence_categories = [\"experimental\",\"author_statement\",\"curator_statement\"]\n",
    "df_go[\"high_confidence\"] = df_go[\"evidence_code\"].apply(lambda x: EVIDENCE_CODES[x] in high_confidence_categories)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Subset to only include the high-quality GO annotations and ouptut the dataset to a csv file.\n",
    "df_go_high_confidence = df_go[df_go[\"high_confidence\"]==True]\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_curated_go_annotations.csv\")\n",
    "df_go_high_confidence = df_go_high_confidence[reshaped_columns]\n",
    "df_go_high_confidence.to_csv(path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>other_gene_identifiers</th>\n",
       "      <th>gene_models</th>\n",
       "      <th>descriptions</th>\n",
       "      <th>annotations</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0044212</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0003700</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0003677</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0016021</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0044212</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0016125</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0032541</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0016125</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0016125</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species unique_gene_identifiers other_gene_identifiers gene_models descriptions annotations sources\n",
       "0      ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "1      ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "2      ath               AT1G01010                          AT1G01010               GO:0044212    TAIR\n",
       "3      ath               AT1G01010                          AT1G01010               GO:0005634    TAIR\n",
       "4      ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "5      ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "6      ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "7      ath               AT1G01010                          AT1G01010               GO:0005634    TAIR\n",
       "8      ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "9      ath               AT1G01010                          AT1G01010               GO:0003700    TAIR\n",
       "10     ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "11     ath               AT1G01010                          AT1G01010               GO:0003677    TAIR\n",
       "12     ath               AT1G01010                          AT1G01010               GO:0016021    TAIR\n",
       "13     ath               AT1G01010                          AT1G01010               GO:0044212    TAIR\n",
       "14     ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "15     ath               AT1G01020                          AT1G01020               GO:0016125    TAIR\n",
       "16     ath               AT1G01020                          AT1G01020               GO:0032541    TAIR\n",
       "17     ath               AT1G01020                          AT1G01020               GO:0016125    TAIR\n",
       "18     ath               AT1G01020                          AT1G01020               GO:0005739    TAIR\n",
       "19     ath               AT1G01020                          AT1G01020               GO:0016125    TAIR"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Outputting the dataset of annotations to a csv file.\n",
    "df_go = df_go[reshaped_columns]\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_all_go_annotations.csv\")\n",
    "df_go.to_csv(path, index=False)\n",
    "df_go.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with plant ontology term annotations (po_[termporal/anatomy]_gene_arabidopsis_tair.assoc)<br>\n",
    "There are two separate files available that include annotations of PO terms. The files do not have headers so column names are added based on how the columns are described in the accompanying available readme files. One of the files contains annotations for PO terms that are spatial, or describe a specific part of plant anatomy or plant molecular structures. The other file contains annotations for PO terms that are temporal, or refer to a specific process or stage of development. These files are each read in separately, and the next cells look at the quantity of unique values in the columns of each dataset. There are more spatial annotations than temporal annotations, and a greater number of terms used to describe the spatial annotations.<br>\n",
    "<br>\n",
    "The next field combines the two datasets of PO annotations and looks at the number of unique values for each column in the resulting dataset. Because there is no overlap in the terms between the two, the datasets are simply appended to one another and the total unique terms are a sum of the individual datasets.<br>\n",
    "<br>\n",
    "Each term annotation in this dataset is also associated with an evidence code specifying the method by which this annotation was made, which is related to the confidence that we can have in this annotation, and the tasks that the annotation should be used for. Almost all of the PO term annotations are high confidence, they are experimentally validated, and only a few of them are derived from author statements.<br>\n",
    "<br>\n",
    "The strings which are described in the synonyms column are included as references to each gene, and are combined with the gene name mentioned in the symbol column into a single bar delimited list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(371438, 7)\n",
      "symbol               24085\n",
      "term_id                369\n",
      "references            2730\n",
      "evidence_code            8\n",
      "name                 25422\n",
      "synonyms             28093\n",
      "type                     6\n"
     ]
    }
   ],
   "source": [
    "#Reading in the dataset of spatial PO term annotations.\n",
    "filename = \"../databases/tair/po_anatomy_gene_arabidopsis_tair.assoc\"\n",
    "df_po_spatial = pd.read_table(filename, header=None, skiprows=0, usecols=[2,4,5,6,9,10,11])\n",
    "df_po_spatial.columns = [\"symbol\",\"term_id\",\"references\",\"evidence_code\",\"name\",\"synonyms\",\"type\"]\n",
    "unique_values = {col:len(pd.unique(df_po_spatial[col].values)) for col in df_po_spatial.columns}\n",
    "print(df_po_spatial.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(186819, 7)\n",
      "symbol               18006\n",
      "term_id                 91\n",
      "references             478\n",
      "evidence_code            4\n",
      "name                 18523\n",
      "synonyms             18973\n",
      "type                     5\n"
     ]
    }
   ],
   "source": [
    "#Reading in the dataset of temporal PO term annotations.\n",
    "filename = \"../databases/tair/po_temporal_gene_arabidopsis_tair.assoc\"\n",
    "df_po_temporal = pd.read_table(filename, header=None, skiprows=0, usecols=[2,4,5,6,9,10,11])\n",
    "df_po_temporal.columns = [\"symbol\",\"term_id\",\"references\",\"evidence_code\",\"name\",\"synonyms\",\"type\"]\n",
    "unique_values = {col:len(pd.unique(df_po_temporal[col].values)) for col in df_po_temporal.columns}\n",
    "print(df_po_temporal.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   symbol                                           synonyms evidence_code\n",
      "0  ATARCA  AT1G18080|ATARCA|RACK1A_AT|RACK1A|SAC53|AtRACK...           IEP\n",
      "1    HMG1  AT1G76490|HMG1|HMGR1|AtHMGR1|MAD3|hydroxy meth...           TAS\n",
      "2    BAS1  AT2G26710|BAS1|CYP734A1|CYP72B1|PHYB ACTIVATIO...           TAS\n",
      "3    BAK1  AT4G33430|BAK1|RKS10|SERK3|ELG|ATSERK3|ATBAK1|...           TAS\n",
      "4    BRI1  AT4G39400|BRI1|CBB2|DWF2|BIN1|ATBRI1|BRASSINOS...           TAS\n",
      "5    ATJ6    AT5G06910|ATJ6|J-domain protein 6|MOJ9.8|MOJ9_8           IEP\n",
      "6    XPB2  AT5G41360|XPB2|ATXPB2|homolog of Xeroderma pig...           NAS\n",
      "7    XPB1  AT5G41370|XPB1|ATXPB1|homolog of xeroderma pig...           NAS\n",
      "8   PPCK2  AT3G04530|PPCK2|PEPCK2|ATPPCK2|phosphoenolpyru...           IEP\n",
      "9    MYC2  AT1G32640|ATMYC2|RD22BP1|JAI1|JIN1|MYC2|ZBF1|J...           NAS\n",
      "(558257, 7)\n",
      "symbol               24148\n",
      "term_id                460\n",
      "references            2807\n",
      "evidence_code            8\n",
      "name                 25506\n",
      "synonyms             28238\n",
      "type                     7\n"
     ]
    }
   ],
   "source": [
    "#Looking at how many unique values each column has.\n",
    "df_po = df_po_spatial.append(df_po_temporal, ignore_index=True)\n",
    "unique_values = {col:len(pd.unique(df_po[col].values)) for col in df_po.columns}\n",
    "print(df_po[[\"symbol\",\"synonyms\",\"evidence_code\"]].head(10))\n",
    "print(df_po.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "experimental               553612\n",
      "curator_statement               0\n",
      "author_statement              118\n",
      "phylogenetics                   0\n",
      "high_throughput              4527\n",
      "computational                   0\n",
      "electronic_annotation           0\n"
     ]
    }
   ],
   "source": [
    "#Quantifying the number of annotations of each type.\n",
    "code_quantities = {c:len([x for x in df_po[\"evidence_code\"] if EVIDENCE_CODES[x] in c]) \n",
    "             for c in list(set(EVIDENCE_CODES.values()))}\n",
    "for k,v in code_quantities.items():\n",
    "    print(\"{:25}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "symbol           object\n",
       "term_id          object\n",
       "references       object\n",
       "evidence_code    object\n",
       "name             object\n",
       "synonyms         object\n",
       "type             object\n",
       "dtype: object"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_po['name'] = df_po['name'].astype(str)\n",
    "df_po.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Restructuring the dataset to include all the expected column names.\n",
    "combine_columns = lambda row, columns: concatenate_with_delim(\"|\", [row[column] for column in columns])\n",
    "df_po[\"species\"] = \"ath\"\n",
    "df_po[\"unique_gene_identifiers\"] = df_po.apply(lambda x: combine_columns(x, [\"symbol\", \"name\"]), axis=1)\n",
    "df_po[\"other_gene_identifiers\"] = df_po[\"synonyms\"]\n",
    "df_po[\"gene_model_strings_1\"] =  df_po[\"unique_gene_identifiers\"].map(lambda x: \"|\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df_po[\"gene_model_strings_2\"] =  df_po[\"unique_gene_identifiers\"].map(lambda x: \"|\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df_po[\"gene_models\"] = df_po.apply(lambda x: combine_columns(x, [\"gene_model_strings_1\", \"gene_model_strings_2\"]), axis=1)\n",
    "df_po[\"descriptions\"] = \"\"\n",
    "df_po[\"annotations\"] = df_po[\"term_id\"]\n",
    "df_po[\"sources\"] = \"TAIR\"\n",
    "df_po[\"other_gene_identifiers\"] = df_po.apply(lambda row: subtract_string_lists(\"|\", row[\"other_gene_identifiers\"],row[\"unique_gene_identifiers\"]), axis=1)\n",
    "df_po = df_po[reshaped_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>other_gene_identifiers</th>\n",
       "      <th>gene_models</th>\n",
       "      <th>descriptions</th>\n",
       "      <th>annotations</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATARCA|AT1G18080</td>\n",
       "      <td>RACK1A_AT|RACK1A|SAC53|AtRACK1|RECEPTOR FOR AC...</td>\n",
       "      <td>AT1G18080</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>HMG1|AT1G76490</td>\n",
       "      <td>HMGR1|AtHMGR1|MAD3|hydroxy methylglutaryl CoA ...</td>\n",
       "      <td>AT1G76490</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>BAS1|AT2G26710</td>\n",
       "      <td>CYP734A1|CYP72B1|PHYB ACTIVATION TAGGED SUPPRE...</td>\n",
       "      <td>AT2G26710</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>BAK1|AT4G33430</td>\n",
       "      <td>RKS10|SERK3|ELG|ATSERK3|ATBAK1|BRI1-associated...</td>\n",
       "      <td>AT4G33430</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>BRI1|AT4G39400</td>\n",
       "      <td>CBB2|DWF2|BIN1|ATBRI1|BRASSINOSTEROID INSENSIT...</td>\n",
       "      <td>AT4G39400</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATJ6|AT5G06910</td>\n",
       "      <td>J-domain protein 6|MOJ9.8|MOJ9_8</td>\n",
       "      <td>AT5G06910</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>XPB2|AT5G41360</td>\n",
       "      <td>ATXPB2|homolog of Xeroderma pigmentosum comple...</td>\n",
       "      <td>AT5G41360</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>XPB1|AT5G41370</td>\n",
       "      <td>ATXPB1|homolog of xeroderma pigmentosum comple...</td>\n",
       "      <td>AT5G41370</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>PPCK2|AT3G04530</td>\n",
       "      <td>PEPCK2|ATPPCK2|phosphoenolpyruvate carboxylase...</td>\n",
       "      <td>AT3G04530</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000013</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>MYC2|AT1G32640</td>\n",
       "      <td>ATMYC2|RD22BP1|JAI1|JIN1|ZBF1|JASMONATE INSENS...</td>\n",
       "      <td>AT1G32640</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000017</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATSBT5.2|AT1G20160</td>\n",
       "      <td>CRSP|SBT5.2(a)|SBT5.2(b)|CO2 RESPONSE SECRETED...</td>\n",
       "      <td>AT1G20160</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000036</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT3.3|AT1G32960</td>\n",
       "      <td>ATSBT3.3|F9L11.13|F9L11_13</td>\n",
       "      <td>AT1G32960</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000036</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT3.12|AT4G21326</td>\n",
       "      <td>ATSBT3.12|subtilase 3.12</td>\n",
       "      <td>AT4G21326</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000036</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT1.3|AT5G51750</td>\n",
       "      <td>ATSBT1.3|subtilase 1.3|MIO24.12|MIO24_12</td>\n",
       "      <td>AT5G51750</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000036</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT1.3|AT5G51750</td>\n",
       "      <td>ATSBT1.3|subtilase 1.3|MIO24.12|MIO24_12</td>\n",
       "      <td>AT5G51750</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000037</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>SEP3|AT1G24260</td>\n",
       "      <td>AGL9|SEPALLATA3|AGAMOUS-like 9|F3I6.19|F3I6_19...</td>\n",
       "      <td>AT1G24260</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000229</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>RABA1b|AT1G16920</td>\n",
       "      <td>RAB11|ATRABA1B|BEX5|RAB GTPase homolog A1B|BFA...</td>\n",
       "      <td>AT1G16920</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000256</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>MYC2|AT1G32640</td>\n",
       "      <td>ATMYC2|RD22BP1|JAI1|JIN1|ZBF1|JASMONATE INSENS...</td>\n",
       "      <td>AT1G32640</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000256</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>RLP4|AT1G28340</td>\n",
       "      <td>AtRLP4|LLR2|receptor like protein 4|F3M18.23|F...</td>\n",
       "      <td>AT1G28340</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000282</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>RLP4|AT1G28340</td>\n",
       "      <td>AtRLP4|LLR2|receptor like protein 4|F3M18.23|F...</td>\n",
       "      <td>AT1G28340</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT3.3|AT1G32960</td>\n",
       "      <td>ATSBT3.3|F9L11.13|F9L11_13</td>\n",
       "      <td>AT1G32960</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ath</td>\n",
       "      <td>EXPA1|AT1G69530</td>\n",
       "      <td>ATEXPA1|EXP1|AT-EXP1|ATEXP1|ATHEXP ALPHA 1.2|e...</td>\n",
       "      <td>AT1G69530</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT1.3|AT5G51750</td>\n",
       "      <td>ATSBT1.3|subtilase 1.3|MIO24.12|MIO24_12</td>\n",
       "      <td>AT5G51750</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT4.12|AT5G59090</td>\n",
       "      <td>ATSBT4.12|subtilase 4.12|K18B18.5|K18B18_5</td>\n",
       "      <td>AT5G59090</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT4.13|AT5G59120</td>\n",
       "      <td>ATSBT4.13|subtilase 4.13|MNC17.1</td>\n",
       "      <td>AT5G59120</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT5.4|AT5G59810</td>\n",
       "      <td>ATSBT5.4|MMN10.6|MMN10_6</td>\n",
       "      <td>AT5G59810</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ath</td>\n",
       "      <td>UBP19|AT2G24640</td>\n",
       "      <td>ubiquitin-specific protease 19|F25P17.6|F25P17_6</td>\n",
       "      <td>AT2G24640</td>\n",
       "      <td></td>\n",
       "      <td>PO:0002000</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATSBT5.2|AT1G20160</td>\n",
       "      <td>CRSP|SBT5.2(a)|SBT5.2(b)|CO2 RESPONSE SECRETED...</td>\n",
       "      <td>AT1G20160</td>\n",
       "      <td></td>\n",
       "      <td>PO:0003011</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT3.3|AT1G32960</td>\n",
       "      <td>ATSBT3.3|F9L11.13|F9L11_13</td>\n",
       "      <td>AT1G32960</td>\n",
       "      <td></td>\n",
       "      <td>PO:0003011</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT4.12|AT5G59090</td>\n",
       "      <td>ATSBT4.12|subtilase 4.12|K18B18.5|K18B18_5</td>\n",
       "      <td>AT5G59090</td>\n",
       "      <td></td>\n",
       "      <td>PO:0003011</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species unique_gene_identifiers                             other_gene_identifiers gene_models descriptions annotations sources\n",
       "0      ath        ATARCA|AT1G18080  RACK1A_AT|RACK1A|SAC53|AtRACK1|RECEPTOR FOR AC...   AT1G18080               PO:0000003    TAIR\n",
       "1      ath          HMG1|AT1G76490  HMGR1|AtHMGR1|MAD3|hydroxy methylglutaryl CoA ...   AT1G76490               PO:0000003    TAIR\n",
       "2      ath          BAS1|AT2G26710  CYP734A1|CYP72B1|PHYB ACTIVATION TAGGED SUPPRE...   AT2G26710               PO:0000003    TAIR\n",
       "3      ath          BAK1|AT4G33430  RKS10|SERK3|ELG|ATSERK3|ATBAK1|BRI1-associated...   AT4G33430               PO:0000003    TAIR\n",
       "4      ath          BRI1|AT4G39400  CBB2|DWF2|BIN1|ATBRI1|BRASSINOSTEROID INSENSIT...   AT4G39400               PO:0000003    TAIR\n",
       "5      ath          ATJ6|AT5G06910                   J-domain protein 6|MOJ9.8|MOJ9_8   AT5G06910               PO:0000003    TAIR\n",
       "6      ath          XPB2|AT5G41360  ATXPB2|homolog of Xeroderma pigmentosum comple...   AT5G41360               PO:0000003    TAIR\n",
       "7      ath          XPB1|AT5G41370  ATXPB1|homolog of xeroderma pigmentosum comple...   AT5G41370               PO:0000003    TAIR\n",
       "8      ath         PPCK2|AT3G04530  PEPCK2|ATPPCK2|phosphoenolpyruvate carboxylase...   AT3G04530               PO:0000013    TAIR\n",
       "9      ath          MYC2|AT1G32640  ATMYC2|RD22BP1|JAI1|JIN1|ZBF1|JASMONATE INSENS...   AT1G32640               PO:0000017    TAIR\n",
       "10     ath      ATSBT5.2|AT1G20160  CRSP|SBT5.2(a)|SBT5.2(b)|CO2 RESPONSE SECRETED...   AT1G20160               PO:0000036    TAIR\n",
       "11     ath        SBT3.3|AT1G32960                         ATSBT3.3|F9L11.13|F9L11_13   AT1G32960               PO:0000036    TAIR\n",
       "12     ath       SBT3.12|AT4G21326                           ATSBT3.12|subtilase 3.12   AT4G21326               PO:0000036    TAIR\n",
       "13     ath        SBT1.3|AT5G51750           ATSBT1.3|subtilase 1.3|MIO24.12|MIO24_12   AT5G51750               PO:0000036    TAIR\n",
       "14     ath        SBT1.3|AT5G51750           ATSBT1.3|subtilase 1.3|MIO24.12|MIO24_12   AT5G51750               PO:0000037    TAIR\n",
       "15     ath          SEP3|AT1G24260  AGL9|SEPALLATA3|AGAMOUS-like 9|F3I6.19|F3I6_19...   AT1G24260               PO:0000229    TAIR\n",
       "16     ath        RABA1b|AT1G16920  RAB11|ATRABA1B|BEX5|RAB GTPase homolog A1B|BFA...   AT1G16920               PO:0000256    TAIR\n",
       "17     ath          MYC2|AT1G32640  ATMYC2|RD22BP1|JAI1|JIN1|ZBF1|JASMONATE INSENS...   AT1G32640               PO:0000256    TAIR\n",
       "18     ath          RLP4|AT1G28340  AtRLP4|LLR2|receptor like protein 4|F3M18.23|F...   AT1G28340               PO:0000282    TAIR\n",
       "19     ath          RLP4|AT1G28340  AtRLP4|LLR2|receptor like protein 4|F3M18.23|F...   AT1G28340               PO:0000293    TAIR\n",
       "20     ath        SBT3.3|AT1G32960                         ATSBT3.3|F9L11.13|F9L11_13   AT1G32960               PO:0000293    TAIR\n",
       "21     ath         EXPA1|AT1G69530  ATEXPA1|EXP1|AT-EXP1|ATEXP1|ATHEXP ALPHA 1.2|e...   AT1G69530               PO:0000293    TAIR\n",
       "22     ath        SBT1.3|AT5G51750           ATSBT1.3|subtilase 1.3|MIO24.12|MIO24_12   AT5G51750               PO:0000293    TAIR\n",
       "23     ath       SBT4.12|AT5G59090         ATSBT4.12|subtilase 4.12|K18B18.5|K18B18_5   AT5G59090               PO:0000293    TAIR\n",
       "24     ath       SBT4.13|AT5G59120                   ATSBT4.13|subtilase 4.13|MNC17.1   AT5G59120               PO:0000293    TAIR\n",
       "25     ath        SBT5.4|AT5G59810                           ATSBT5.4|MMN10.6|MMN10_6   AT5G59810               PO:0000293    TAIR\n",
       "26     ath         UBP19|AT2G24640   ubiquitin-specific protease 19|F25P17.6|F25P17_6   AT2G24640               PO:0002000    TAIR\n",
       "27     ath      ATSBT5.2|AT1G20160  CRSP|SBT5.2(a)|SBT5.2(b)|CO2 RESPONSE SECRETED...   AT1G20160               PO:0003011    TAIR\n",
       "28     ath        SBT3.3|AT1G32960                         ATSBT3.3|F9L11.13|F9L11_13   AT1G32960               PO:0003011    TAIR\n",
       "29     ath       SBT4.12|AT5G59090         ATSBT4.12|subtilase 4.12|K18B18.5|K18B18_5   AT5G59090               PO:0003011    TAIR"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Outputting the dataset of annotations to a csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_curated_po_annotations.csv\")\n",
    "df_po.to_csv(path, index=False)\n",
    "df_po.head(30)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
